File: Python Snippets/rpn_calculator.py
def divisionWithTest(dividend, divisor):
  if divisor == 0:
    return None
  else:
    return dividend / divisor

def expression(lhs, rhs, operation):
  if operation == "+":
    return lhs + rhs
  elif operation == "-":
    return lhs - rhs
  elif operation == "*":
    return lhs * rhs
  elif operation == "/":
    return divisionWithTest(lhs, rhs)

def reversePolishCalc(expressionText):
  '''
  receives math expression as a string written in
  reverse polish nonation and returns the result
  '''
  lhs = rhs = result = None

  for char in expressionText.split():
    if char.isnumeric():
      if not lhs:
        lhs = int(char)
      else:
        rhs = int(char)
    elif char in "+-*/" and rhs != None:
      # "pushes" the result to left-hand side so following
      # operations are done with it
      lhs = result = expression(lhs, rhs, char)
      rhs = None

  return result

File: R Snippets/rpn_calculator.r
expression <- function(lhs, rhs, operator) {
  if (operator == "+") {
    return (lhs + rhs)
  } else if (operator == "-") {
    return (lhs - rhs)
  } else if (operator == "*") {
    return (lhs * rhs)
  } else if (operator == "/") {
    return (division_with_test(lhs, rhs))
  }
}

division_with_test <- function(dividend, divisor) {
  if (divisor == 0) {
    return (NA)
  }
  return (dividend / divisor)
}

polish_calc <- function(calc_text) {
  split_text <-  strsplit(calc_text, " ")[[1]]
  lhs <- NULL
  rhs <- NULL
  operator <- NULL
  result <- NULL

  for (char in split_text) {
    if (grepl(char, "+-*/")) {
      # print(paste("The operator is now:", char))
      operator = char
    }

    is_complete = !is.null(lhs) && !is.null(rhs) && !is.null(operator)
    if (is_complete) {
      result = expression(lhs, rhs, operator)
      lhs = result
      rhs = NULL
      operator = NULL
      next
    }

    if (!grepl("[0-9]", char)) {
      # print(paste("Char:", char, "is not numeric."))
      next
    } else if (is.null(lhs)) {
      lhs = as.integer(char)
      # print(paste("The lhs is now:", char))
      next
    } else {
      rhs = as.integer(char)
      # print(paste("The rhs is now:", char))
      next
    }
  }

  return (result)
}

expression_text = "10 10 *"
polish_calc(expression_text)
